nixos/modules/services/web-servers/garage.nix

```diff
@@ -125,18 +125,25 @@ in
       restartTriggers = [
         configFile
       ] ++ (lib.optional (cfg.environmentFile != null) cfg.environmentFile);
-      serviceConfig = {
-        ExecStart = "${cfg.package}/bin/garage server";
-
-        StateDirectory = mkIf (
-          anyHasPrefix "/var/lib/garage" cfg.settings.data_dir
-          || hasPrefix "/var/lib/garage" cfg.settings.metadata_dir
-        ) "garage";
-        DynamicUser = lib.mkDefault true;
-        ProtectHome = true;
-        NoNewPrivileges = true;
-        EnvironmentFile = lib.optional (cfg.environmentFile != null) cfg.environmentFile;
-      };
+      serviceConfig =
+        let
+          paths = lib.flatten [
+            cfg.settings.data_dir
+            cfg.settings.metadata_dir
+          ];
+          isDefaultStateDirectory = lib.any (lib.hasPrefix "/var/lib/garage") paths;
+          isDefaultReadWritePaths = lib.all (lib.hasPrefix "/var/lib/garage") paths;
+        in
+        {
+          ExecStart = "${cfg.package}/bin/garage server";
+
+          StateDirectory = lib.mkIf isDefaultStateDirectory "garage";
+          DynamicUser = lib.mkDefault true;
+          ProtectHome = true;
+          NoNewPrivileges = true;
+          EnvironmentFile = lib.optional (cfg.environmentFile != null) cfg.environmentFile;
+          ReadWritePaths = lib.optional (!isDefaultReadWritePaths) paths;
```

Comment:
I think this should filter out any entries in `paths` that are under `/var/lib/garage` so they are not duplicated here.